crates/binjs_io/src/entropy/model.rs

```diff
@@ -1,183 +1,275 @@
-use entropy::{ DecodingModel, EncodingModel, Model };
-use entropy::predict::{ ContextPredict, PathPredict, Symbol };
-use entropy::tree::{ EXPECTED_PATH_DEPTH, EXPECTED_SCOPE_DEPTH, ASTPath, F64, Label, ScopeIndex, ScopePath, SharedTree, Visitor, WalkTree };
+// use entropy::predict::{ ContextPredict, PathPredict, Symbol };
 
-use binjs_shared::{ IdentifierName, InterfaceName, SharedString };
-use binjs_shared::ast::PathItem;
+use entropy::predict::PathPredict;
+
+use io::TokenWriter;
+use ::TokenWriterError;
+
+use binjs_shared::{ F64, FieldName, IdentifierName, InterfaceName, PropertyKey, SharedString };
 
 use std;
-use std::rc::Rc;
+use std::collections::HashMap;
+
+pub type IOPath = binjs_shared::ast::Path<InterfaceName, (/* child index */ usize, /* field name */ FieldName)>;
+
+#[derive(Debug, Default)]
+pub struct Dictionary<T> {
+    /// All booleans appearing in the AST, predicted by path.
+    pub bool_by_path: PathPredict<Option<bool>, T>,
+
+    /// All floats appearing in the AST, predicted by path.
+    pub float_by_path: PathPredict<Option<F64>, T>,
+
+    /// All unsigned longs appearing in the AST, predicted by path.
+    pub unsigned_long_by_path: PathPredict<u32, T>,
+
+    /// All string enumerations, predicted by path.
+    pub string_enum_by_path: PathPredict<SharedString, T>,
+
+    /// All property keys, predicted by path.
+    pub property_key_by_path: PathPredict<Option<PropertyKey>, T>,
+
+    /// All identifier names, predicted by path.
+    pub identifier_name_by_path: PathPredict<Option<IdentifierName>, T>,
+
+    /// All interface names, predicted by path.
+    pub interface_name_by_path: PathPredict<InterfaceName, T>,
 
-pub struct ExactModel;
+    /// All string literals, predicted by path.
+    pub string_literal_by_path: PathPredict<Option<SharedString>, T>,
 
-impl Model for ExactModel {
-    fn encoding(&self, tree: &SharedTree) -> Box<EncodingModel> {
-        Box::new(ExactEncodingModel::new(tree))
+    /// All list lengths, predicted by path.
+    pub list_length_by_path: PathPredict<Option<u32>, T>,
+
+    // Missing:
+    // - offsets (cannot be predicted?)
+    // - property keys predicted by window?
+    // - identifier names predicted by window?
+    // - literal strings by window?
+    // - directives?
+    // - property keys, identifier names, literal strings by file?
+}
+impl<T> Dictionary<T> {
+    /// Return the number of states in this dictionary.
+    pub fn len(&self) -> usize {
+          self.bool_by_path.len()
+        + self.float_by_path.len()
+        + self.unsigned_long_by_path.len()
+        + self.string_enum_by_path.len()
+        + self.property_key_by_path.len()
+        + self.identifier_name_by_path.len()
+        + self.interface_name_by_path.len()
+        + self.string_literal_by_path.len()
+        + self.list_length_by_path.len()
     }
 }
 
+/// Strings grouped per kind.
```

Comment:
to my understanding, this struct is actually "a map from kind+string to something (T)".
that means, "string" is part of key, not value.

and the actual usage is the following 2 cases:
  * T = the number of appearance of kind+string in single file
  * T = the number of files that has kind+string, among all files

so, "StringsPerKind" still doesn't reflect the actual meaning.
given fields have "instances", how about "InstancesPerKindAndString" ?

Reply:
What about `KindedStringMap`? That sounds a bit more readable.

Reply:
I'm not sure "Kinded" means what we need, but if it means, I think it's good.

Reply:
Well, I think it can be accepted with that meaning :)